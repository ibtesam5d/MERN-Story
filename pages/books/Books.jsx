import React, { useState } from "react";
import "./Books.scss";
import { BsChevronDown } from "react-icons/bs";
import { books } from "../../data";
import BookCard from "../../src/components/BookCard/BookCard";

const Books = () => {
  const [open, setOpen] = useState(false);

  const [sort, setSort] = useState("sales");

  const reSort = (type) => {
    setSort(type);
    setOpen(false);
  };

  return (
    <div className="books">
      <div className="container">
        <span className="breadcrumbs">Story - Action</span>
        <h1>Non-stop Action</h1>
        <p>
          Explore the world of fights and dramas with story's action library
        </p>
        <div className="menu">
          <div className="left">
            <span>Price</span>
            <input type="text" placeholder="min" />
            <input type="text" placeholder="max" />
            <button>Apply</button>
          </div>
          <div className="right">
            <span className="sortBy">Sortby</span>
            <span className="sortType">
              {sort === "sales" ? "Best Selling" : "Newest"}
            </span>
            <BsChevronDown className="down" onClick={() => setOpen(!open)} />

            {open && (
              <div className="rightMenu">
                {sort === "sales" ? (
                  <span onClick={() => reSort("createdAt")}>Newest</span>
                ) : (
                  <span onClick={() => reSort("sales")}>Best Selling</span>
                )}
              </div>
            )}
          </div>
        </div>

        {/* ====== BOOKS ======== */}
        <div className="cards">
          {books.map((book) => (
            <BookCard item={book} key={book.id} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default Books;
