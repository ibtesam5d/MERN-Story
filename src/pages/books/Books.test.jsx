import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import { expect, it } from "vitest";
import { MemoryRouter } from "react-router-dom";
import Books from "./Books";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

it("should show h1 tag", async () => {
  const queryClient = new QueryClient();
  const route = "/books?category=all";
  render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter initialEntries={[route]}>
        <Books />
      </MemoryRouter>
    </QueryClientProvider>
  );
  const titleTag = screen.getByTestId("h1");
  expect(titleTag).toBeVisible();
});

it("should show bookCard component", async () => {
  const queryClient = new QueryClient();
  const route = "/books?category=all";
  render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter initialEntries={[route]}>
        <Books />
      </MemoryRouter>
    </QueryClientProvider>
  );
  const bookCard = screen.getByTestId("book-card");
  expect(bookCard).toBeVisible();
});

it("should show both input", async () => {
  const queryClient = new QueryClient();
  const route = "/books?category=all";
  render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter initialEntries={[route]}>
        <Books />
      </MemoryRouter>
    </QueryClientProvider>
  );
  const minInput = screen.getByTestId("min");
  const maxInput = screen.getByTestId("max");
  expect(minInput).toBeVisible();
  expect(maxInput).toBeVisible();
});

it("should update both input's value on change", async () => {
  const queryClient = new QueryClient();
  const route = "/books?category=all";
  render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter initialEntries={[route]}>
        <Books />
      </MemoryRouter>
    </QueryClientProvider>
  );
  const minInput = screen.getByTestId("min");
  const maxInput = screen.getByTestId("max");
  const testValue = "10";
  fireEvent.change(minInput, { target: { value: testValue } });
  fireEvent.change(maxInput, { target: { value: testValue } });
  expect(minInput.value).toBe(testValue);
  expect(maxInput.value).toBe(testValue);
});

it("matches snapshot of the Books page", async () => {
  const queryClient = new QueryClient();
  const route = "/books?category=all";
  const wrapperEl = render(
    <QueryClientProvider client={queryClient}>
      <MemoryRouter initialEntries={[route]}>
        <Books />
      </MemoryRouter>
    </QueryClientProvider>
  );
  expect(wrapperEl).toMatchSnapshot();
});
