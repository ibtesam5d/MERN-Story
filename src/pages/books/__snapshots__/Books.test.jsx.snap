// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot of the Books page 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="books"
        title="books"
      >
        <div
          class="container"
        >
          <span
            class="breadcrumbs"
          >
            Story - all
          </span>
          <h1
            data-testid="h1"
          >
            all
          </h1>
          <p>
            Explore the world of fights and dramas with story's rich library
          </p>
          <div
            class="menu"
          >
            <div
              class="left"
            >
              <span>
                Price
              </span>
              <input
                data-testid="min"
                placeholder="min"
                type="text"
              />
              <input
                data-testid="max"
                placeholder="max"
                type="text"
              />
              <button>
                Apply
              </button>
            </div>
            <div
              class="right"
            >
              <span
                class="sortBy"
              >
                Sortby
              </span>
              <span
                class="sortType"
              >
                Best Selling
              </span>
              <svg
                class="down"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
          <div
            class="cards"
            data-testid="book-card"
          >
            Loading..
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="books"
      title="books"
    >
      <div
        class="container"
      >
        <span
          class="breadcrumbs"
        >
          Story - all
        </span>
        <h1
          data-testid="h1"
        >
          all
        </h1>
        <p>
          Explore the world of fights and dramas with story's rich library
        </p>
        <div
          class="menu"
        >
          <div
            class="left"
          >
            <span>
              Price
            </span>
            <input
              data-testid="min"
              placeholder="min"
              type="text"
            />
            <input
              data-testid="max"
              placeholder="max"
              type="text"
            />
            <button>
              Apply
            </button>
          </div>
          <div
            class="right"
          >
            <span
              class="sortBy"
            >
              Sortby
            </span>
            <span
              class="sortType"
            >
              Best Selling
            </span>
            <svg
              class="down"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="cards"
          data-testid="book-card"
        >
          Loading..
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`should have img tag 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="books"
        title="books"
      >
        <div
          class="container"
        >
          <span
            class="breadcrumbs"
          >
            Story - all
          </span>
          <h1
            data-testid="h1"
          >
            all
          </h1>
          <p>
            Explore the world of fights and dramas with story's rich library
          </p>
          <div
            class="menu"
          >
            <div
              class="left"
            >
              <span>
                Price
              </span>
              <input
                data-testid="min"
                placeholder="min"
                type="text"
              />
              <input
                data-testid="max"
                placeholder="max"
                type="text"
              />
              <button>
                Apply
              </button>
            </div>
            <div
              class="right"
            >
              <span
                class="sortBy"
              >
                Sortby
              </span>
              <span
                class="sortType"
              >
                Best Selling
              </span>
              <svg
                class="down"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
          <div
            class="cards"
            data-testid="book-card"
          >
            Loading..
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="books"
      title="books"
    >
      <div
        class="container"
      >
        <span
          class="breadcrumbs"
        >
          Story - all
        </span>
        <h1
          data-testid="h1"
        >
          all
        </h1>
        <p>
          Explore the world of fights and dramas with story's rich library
        </p>
        <div
          class="menu"
        >
          <div
            class="left"
          >
            <span>
              Price
            </span>
            <input
              data-testid="min"
              placeholder="min"
              type="text"
            />
            <input
              data-testid="max"
              placeholder="max"
              type="text"
            />
            <button>
              Apply
            </button>
          </div>
          <div
            class="right"
          >
            <span
              class="sortBy"
            >
              Sortby
            </span>
            <span
              class="sortType"
            >
              Best Selling
            </span>
            <svg
              class="down"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 16 16"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z"
                fill-rule="evenodd"
              />
            </svg>
          </div>
        </div>
        <div
          class="cards"
          data-testid="book-card"
        >
          Loading..
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
